ankidroid/Anki-Android
# Add Crop functionality

## Pull Request template

## Purpose / Description

There was no way to crop images previously, but cameras especially are generating very large images these days

This allows the advanced editor's image field component to crop images, and prompts users to do so for photos or if the image is large

This is entirely based on the work from @NightXlt - everything about how to do the Intent correctly, to the URI / File path conversion is their work

I took the PR contents of #5301 and re-factored them to fit better with the work done since @NightXlt originally submitted the work, then I tested, tested, and tested some more with little fixes related to preview handling and button visibility added in.

Finally, I deconstructed the work into chunks as small as I could so review would be easier.

**Please understand the commit separation was done after the fact and there may be a couple cross-dependencies still present in the commit**. The separate commits are for easier review but I make no guarantee they will function separately, testing has only been performed with the complete work


## Fixes
Fixes #2554 
Obsoletes #5301 

## Approach

The basic idea is to 

- handle the path internally with a path backup
- add the general ability to create temporary files
- add the general ability to generate URIs for our cache files
- add the general ability to translate between URIs and Paths
- ask the system to crop when we want, with a URI to a new internal file for the result



## How Has This Been Tested?

API16, 18, 24, 29, 30 emulators

The scenarios I ran are:

- add note, paperclip, add image, gallery, verify preview, choose crop, do the crop, verify preview, save, note editor preview
- same as above but camera
- save note, view note verify images

Note that after adding images but before you save the note, the field has data and the paperclip menu behaves differently than if you were editing a note and it had data because the paperclip menu listener hasn't re-been rebound. This is an existing bug and not addressed here

Same as above but with an existing note with image data:
- edit note (with image data), paperclip (opens advanced image editor w/preview), crop, do crop+return, verify image, save, preview from note editor, verify image, return and save note, view note and verify images


## Learning (optional, can help others)

Of interest from this PR:

Attempting to implement an "EDIT" intent vs stricly crop is better but fails in mysterious ways, especially with API30

API30 in general may require more testing

Of interest that existed prior to the PR and still exist:

The advanced editor is rough to deal with, it's pretty fluffy with regard to number of objects and how control passes around

The paperclip menu behaves inconsistently right now based on whether the note had data in the field to start or was empty. If empty to start it will show a popup menu, if it had data to start it sends you right to the advanced editor.

The advanced editor doesn't handle the case of multiple things in the same a field very well.

The advanced editor doesn't handle the case of editing a multimedia field very well (it just keeps appending the edits) but at least with preview from note editor this is easy to fix



## Checklist
_Please, go through these checks before submitting the PR._
